pinot-segment-local/src/main/java/org/apache/pinot/segment/local/recordtransformer/SpecialValueTransformer.java

```diff
@@ -18,112 +18,119 @@
+      Object value = record.getValue(column);
       if (value instanceof Object[]) {
         // Multi-valued column.
         Object[] values = (Object[]) value;
-        int numValues = values.length;
-        List<Object> negativeZeroNanSanitizedValues = new ArrayList<>(numValues);
-        for (Object o : values) {
-          Object zeroTransformedValue = transformNegativeZero(o);
-          Object nanTransformedValue = transformNaN(zeroTransformedValue);
-          if (nanTransformedValue != null) {
-            negativeZeroNanSanitizedValues.add(nanTransformedValue);
+        List<Object> transformedValues = new ArrayList<>(values.length);
+        boolean transformed = false;
+        for (Object v : values) {
+          Object transformedValue = transformValue(v);
+          if (transformedValue != v) {
+            transformed = true;
+          }
+          if (transformedValue != null) {
+            transformedValues.add(transformedValue);
+          }
+          if (transformed) {
+            record.putValue(column, !transformedValues.isEmpty() ? transformedValues.toArray() : null);
           }
         }
-        record.putValue(element, negativeZeroNanSanitizedValues.toArray());
-      } else {
+      } else if (value != null) {
         // Single-valued column.
-        Object zeroTransformedValue = transformNegativeZero(value);
-        Object nanTransformedValue = transformNaN(zeroTransformedValue);
-        if (nanTransformedValue != value) {
-          record.putValue(element, nanTransformedValue);
+        Object transformedValue = transformValue(value);
+        if (transformedValue != value) {
+          record.putValue(column, transformedValue);
         }
       }
     }
-    if (_negativeZeroConversionCount > 0 || _nanConversionCount > 0) {
-      LOGGER.debug("Converted {} -0.0s to 0.0 and {} NaNs to null", _negativeZeroConversionCount, _nanConversionCount);
-    }
     return record;
   }
 
-  @VisibleForTesting
-  int getNegativeZeroConversionCount() {
-    return _negativeZeroConversionCount;
-  }
-
-  @VisibleForTesting
-  int getNanConversionCount() {
-    return _nanConversionCount;
+  @Nullable
+  private Object transformValue(Object value) {
+    if (value instanceof Float) {
+      Float floatValue = (Float) value;
+      if (floatValue.isNaN()) {
+        return null;
+      }
+      if (Float.floatToRawIntBits(floatValue) == NEGATIVE_ZERO_FLOAT_BITS) {
+        return 0.0f;
+      }
+    } else if (value instanceof Double) {
+      Double doubleValue = (Double) value;
+      if (doubleValue.isNaN()) {
+        return null;
+      }
+      if (Double.doubleToRawLongBits(doubleValue) == NEGATIVE_ZERO_DOUBLE_BITS) {
+        return 0.0d;
+      }
+    } else if (value instanceof BigDecimal) {
+      BigDecimal bigDecimalValue = (BigDecimal) value;
+      BigDecimal stripped = bigDecimalValue.stripTrailingZeros();
+      if (!stripped.equals(bigDecimalValue)) {
+        return stripped;
+      }
```

Comment:
doesn't this fundamentally change the data, though. If you have `10.100`, this will return `10.1`. But there's many scientific applications where those 2 trailing zeroes indicate actual significant figures and can't just be removed.

Reply:
I don't fully follow, can you elaborate? We need to strip zeros because `10.100` and `10.1` can also mess up the index.

Reply:
ya, agreed that this messes up the index, but it's a problem with the index rather than the data. `10.100` actually has a different meaning than `10.1`, so we can't just transform one into the other. But `0` vs `0e-18` vs `0e18`, as far as I know, are all just 0.

Reply:
Hmm, big decimal is mostly used to store metric values for lossless aggregate. In case when `10.100` stands for different meaning from `10.1`, what is the actual use case? Does it make sense to store it as STRING?